### PR TITLE
[X86AsmParser] Check displacement overflow

### PR DESCRIPTION
**Pull request by MaskRay**
A displacement is an 8-, 16-, or 32-bit value.
LLVM integrated assembler silently encodes an out-of-range displacement.

GNU assembler checks the displacement and may report a warning or error
(error is for 64-bit addressing, done as part of
https://sourceware.org/PR10636).
```
movq 0x80000000(%rip), %rax
Error: 0x80000000 out of range of signed 32bit displacement

movq -0x080000001(%rax), %rax
Error: 0xffffffff7fffffff out of range of signed 32bit displacement

movl 0x100000001(%eax), %eax
Warning: 0x100000001 shortened to 0x1
```

For 32-bit addressing, GNU assembler gives no diagnostic when the
displacement is within `[-2**32,2**32)`. 16-bit addressing is similar.
```
movl 0xffffffff(%eax), %eax  # no diagnostic
movl -0xffffffff(%eax), %eax # no diagnostic
```

Supporting a larger range is probably because wraparound using a large
constant is more reasonable. E.g. Linux kernel arch/x86/kernel/head_32.S
has `leal -__PAGE_OFFSET(%ecx),%esp` where `__PAGE_OFFSET` is
0xc0000000.

This patch implements a similar behavior.

### CODE REVIEW

File: llvm/test/MC/X86/displacement-overflow.s [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+leal -0xffffffff(%eax), %eax
+
+# CHECK: [[#@LINE+1]]:19: warning: displacement 4294967296 shortened to signed 32-bit
+movl 0xffffffff+1(%eax), %eax
+
+# CHECK: [[#@LINE+1]]:20: warning: displacement -4294967296 shortened to signed 32-bit
+leal -0xffffffff-1(%eax), %eax
+
+{disp8} leal 0x100(%ebx), %eax
+{disp8} leal -0x100(%ebx), %eax
```
Comment (phoebewang) [inline llvm/test/MC/X86/displacement-overflow.s:25]: Should we emit `warning: displacement 256 is not within [-128, 127] for disp8 encoding`?

Comment (MaskRay) [reply llvm/test/MC/X86/displacement-overflow.s:25]: gas doesn't report a diagnostic for `{disp8}` or the deprecated `.d8` suffix.

I think `{disp8}` is to disable `modrm.mod==0` and force a displacement, but doesn't care whether the displacement is out of `int8_t`..

Comment (phoebewang) [reply llvm/test/MC/X86/displacement-overflow.s:25]: I think `{disp8}` is mainly used for AVX512 instructions which have "compressed disp8*N encoding".
I'm ok with this given gas doesn't do it either.

**Review by phoebewang — APPROVED** [review]
LGTM.

File: llvm/lib/Target/X86/AsmParser/X86AsmParser.cpp [hunk 2]
```diff
@@ -3062,6 +3062,33 @@ bool X86AsmParser::ParseMemOperand(unsigned SegReg, const MCExpr *Disp,
...
+      auto Imm = CE->getValue();
+      bool Is64 = X86MCRegisterClasses[X86::GR64RegClassID].contains(BaseReg) ||
+                  X86MCRegisterClasses[X86::GR64RegClassID].contains(IndexReg);
+      bool Is16 = X86MCRegisterClasses[X86::GR16RegClassID].contains(BaseReg);
+      if (Is64) {
+        if (!isInt<32>(Imm))
+          return Error(BaseLoc, "displacement " + Twine(Imm) +
+                                    " is not within [-2147483648, 2147483647]");
+      } else if (!Is16) {
+        if (!isUInt<32>(Imm < 0 ? -uint64_t(Imm) : uint64_t(Imm))) {
```
Comment (KanRobert) [inline llvm/lib/Target/X86/AsmParser/X86AsmParser.cpp:3081]: Could we print the final value too?
https://www.godbolt.org/z/3ev7KPdno

**Review by KanRobert — APPROVED** [review]
LGTM

**Pull request merged**